src/aws-cpp-sdk-core/include/smithy/client/features/ChunkingInterceptor.h

```diff
@@ -0,0 +1,206 @@
+#include <aws/core/utils/Array.h>
+#include <aws/core/utils/StringUtils.h>
+#include <aws/core/utils/HashingUtils.h>
+#include <aws/core/utils/logging/LogMacros.h>
+#include <aws/core/utils/memory/stl/AWSStringStream.h>
+#include <aws/core/utils/memory/stl/AWSVector.h>
+#include <smithy/interceptor/Interceptor.h>
+#include <aws/core/client/ClientConfiguration.h>
+#include <aws/core/utils/Outcome.h>
+#include <aws/core/client/AWSError.h>
+#include <memory>
+
+namespace smithy {
+namespace client {
+namespace features {
+
+static const size_t AWS_DATA_BUFFER_SIZE = 65536;
+static const char* ALLOCATION_TAG = "ChunkingInterceptor";
+static const char* CHECKSUM_HEADER_PREFIX = "x-amz-checksum-";
+
+template <size_t DataBufferSize = AWS_DATA_BUFFER_SIZE>
+class AwsChunkedStreamBuf : public std::streambuf {
+public:
+    AwsChunkedStreamBuf(Aws::Http::HttpRequest* request,
+                        const std::shared_ptr<Aws::IOStream>& stream,
+                        size_t bufferSize = DataBufferSize)
+        : m_request(request),
+          m_stream(stream),
+          m_data(bufferSize)
+    {
+        assert(m_stream != nullptr);
+        if (m_stream == nullptr) {
+            AWS_LOGSTREAM_ERROR("AwsChunkedStream", "stream is null");
+        }
+        assert(m_request != nullptr);
+        if (m_request == nullptr) {
+            AWS_LOGSTREAM_ERROR("AwsChunkedStream", "request is null");
+        }
+    }
+
+protected:
+    int_type underflow() override {
+        if (gptr() && gptr() < egptr()) {
+            return traits_type::to_int_type(*gptr());
+        }
+
+        // only read and write to chunked stream if the underlying stream
+        // is still in a valid state
+        if (m_stream->good()) {
+            // Try to read in a 64K chunk, if we cant we know the stream is over
+            m_stream->read(m_data.GetUnderlyingData(), m_data.GetLength());
+            size_t bytesRead = static_cast<size_t>(m_stream->gcount());
+            writeChunk(bytesRead);
+
+            // if we've read everything from the stream, we want to add the trailer
+            // to the underlying stream
+            if ((m_stream->peek() == EOF || m_stream->eof()) && !m_stream->bad()) {
+                writeTrailerToUnderlyingStream();
+            }
+        }
+
+        // if the chunking buffer is empty there is nothing to read
+        if (m_chunkingBufferPos >= m_chunkingBuffer.size()) {
+            return traits_type::eof();
+        }
+
+        // Set up buffer pointers to read from chunking buffer
+        size_t remainingBytes = m_chunkingBuffer.size() - m_chunkingBufferPos;
+        size_t bytesToRead = std::min(remainingBytes, DataBufferSize);
+        
+        setg(m_chunkingBuffer.data() + m_chunkingBufferPos, 
+             m_chunkingBuffer.data() + m_chunkingBufferPos, 
+             m_chunkingBuffer.data() + m_chunkingBufferPos + bytesToRead);
+        
+        m_chunkingBufferPos += bytesToRead;
```

Comment:
is data ever erased from the vector? otherwise this vector would grow to the size of the body right?